kythe/rust/extractor/src/bin/save_analysis.rs

```diff
@@ -55,5 +57,53 @@ fn generate_arguments(arguments: Vec<String>, output_dir: &Path) -> Result<Vec<S
         output_dir.to_str().ok_or_else(|| anyhow!("Couldn't convert temporary path to string"))?;
     rustc_arguments[outdir_position] = format!("--out-dir={}", outdir_str);
 
+    // Check if there is a file containing environment variables
+    let env_file_position = arguments.iter().position(|arg| arg.contains("--env-file"));
+    if let Some(position) = env_file_position {
+        let env_file = &arguments[position + 1];
+        let path = Path::new(env_file);
+        process_env_file(path).context("Failed to process wrapper environment file")?;
+    }
+
+    // Check if there is a file containing extra compiler arguments
+    let arg_file_position = arguments.iter().position(|arg| arg.contains("--arg-file"));
+    if let Some(position) = arg_file_position {
+        let arg_file = &arguments[position + 1];
+        let path = Path::new(arg_file);
+        let new_arguments =
+            process_arg_file(path).context("Failed to process wrapper argument file")?;
+        rustc_arguments.extend(new_arguments);
+    }
+
     Ok(rustc_arguments)
 }
+
+/// Process an environment variable file passed to the process wrapper
+///
+/// * `path` - The path of the environment variable file
+fn process_env_file(path: &Path) -> Result<()> {
+    let file = File::open(path)?;
+    let lines = io::BufReader::new(file).lines();
+    let pwd = std::env::current_dir().context("Couldn't determine pwd")?;
+    for line in lines {
+        let line_string = line.context("Failed to read line of env file")?;
+        let split: Vec<&str> = line_string.split('=').collect();
+        let value = split[1].replace("${pwd}", pwd.to_str().unwrap());
+        std::env::set_var(split[0], value);
+    }
+    Ok(())
+}
+
+/// Process a compiler argument file passed to the process wrapper
+///
+/// * `path` - The path of the compiler argument file
+fn process_arg_file(path: &Path) -> Result<Vec<String>> {
+    let file = File::open(path)?;
+    let lines = io::BufReader::new(file).lines();
+    let mut arguments = Vec::new();
+    for line in lines {
+        let line_string = line.context("Failed to read line of arg file")?;
+        arguments.push(line_string);
+    }
+    Ok(arguments)
```

Comment:
FWIW, this is also just:
```
    io::BufReader::new(File::open(path)?)
        .lines()
        .map(|line| line.context("..."))
        .collect()
```

Reply:
Oh I actually never knew that a question mark could be used inline. However, the iterator returns an Option so I don't believe that map would work correctly.

Reply:
I stand corrected